### ISSUE TITLE
Windows Support

### ISSUE THREAD

**Issue by programmarchy** [sandeepmistry/noble#84]
Saw the readme, but wanted to bump this in issues to start a discussion:

Anyone currently working on a Windows 8 implementation?

**Comment by sandeepmistry** [comment]
This is a good related read by @punchthrough: [Windows Development – Part One - LightBlue Bean](https://punchthrough.com/bean/windows-development-part-one/)

**Comment by voodootikigod** [comment]
I am going to try to tackle this unless anyone else has already done so and syntehsize/finish out the work already done... If anyone (wink wink @sandeepmistry ) knows of existing starter forks, let me know would rather start with exiting work than blazing trails (NIH). 

Cheers

**Comment by sandeepmistry** [comment]
@mhle did email me last month regards this, we discussed using NodeRT. Not sure if he's made any progress, haven't heard anything back from him. @mhle have you made any progress porting noble to run on Windows 8.1?

@voodootikigod other than that, do not know of any other forks. It would be awesome if you could blaze a trail for this.

There are some noble API's that do not have equivalents in Windows 8.1, like start/stop scanning, connecting/disconnecting. See linked article above for more info, but those need to be initiated via the Windows UI. We can also ask @raykamp for more info. on their experiences with Windows 8.1 and BLE.

**Comment by mhle** [comment]
As Sandeep mentioned I've used NodeRT as a way to access the WinRT API for bluetooth. So far it's worked fine, although you should be aware that even though NodeRT exposes most of the functionality, some are missing and so you cannot always rely on the MSDN docs.

- No promises (there is a way to do this according to NodeRT documentation using nodert-promises but didn't get it to work)
- getting values from WinRT key,value containers are not straightforward

Also same limitations as the ones mentioned in the link above from the Punchthrough blog, with the main ones being:

- No access to RSSI signal strength indication.
- If a device disconnects somehow on its own, it shows up as “Paired” in the list and cannot be re-connected to without removing it and re-pairing.
 -“Paired” devices remain in the list until they’re removed by the user. It isn’t clear whether the device is actually present or not.
- Applications cannot help with connection management in any way.
- Only “Connected” devices are available to the application.
- The user has to manage connection and removal entirely.

In addition there is also currently no way to access advertisement data using the API.

**Comment by sandeepmistry** [comment]
Here are some interesting links on Windows 10, that Peter Vagra emailed me:

http://wmpoweruser.com/windows-10-will-support-ibeacon-technology/
http://rarcher.azurewebsites.net/Post/PostContent/47

```
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisement
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementBytePatten
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementDataSection
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementDataTypes
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementFilter
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementPublisher
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementPublisherStatusChangedEventArgs
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementReceivedEventArgs
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementWatcher
Windows.Devices.Bluetooth.Advertisement.BluetoothLEAdvertisementWatcherStoppedEventArgs
Windows.Devices.Bluetooth.Advertisement.BluetoothLEManufacturerData
Windows.Devices.Bluetooth.Background.BluetoothLEAdvertisementPublisherTriggerDetails
Windows.Devices.Bluetooth.Background.BluetoothLEAdvertisementWatcherTriggerDetails
```

Looks like there will be some API's for scanning and maybe advertising as well. Didn't see anything on connecting.

**Comment by mhle** [comment]
Thanks Sandeep,

Hopefully developing BLE apps on Win10 will be a much better experience. 

Michael

**Comment by programmarchy** [comment]
Agree that Windows 10 will be much better. I'm currently working on a C++ API for BLE on Windows 8.1 and should have some code I can contribute to this project in the next few weeks.

If you're getting started @voodootikigod I found these links to be very helpful for getting up and running:

What libraries you need to link and how to order C headers:
http://social.msdn.microsoft.com/Forums/en-US/1bc0a585-c886-4686-99e1-57d92210c460/how-to-retrive-gatt-attributes-using-bluetooth-le-apis-in-win32-console-application?forum=wdk

Overview of how Windows interfaces with BLE:
http://msdn.microsoft.com/en-us/library/windows/hardware/jj159880(v=vs.85).aspx

Getting a BLE handle from a GUID:
https://social.msdn.microsoft.com/Forums/windowshardware/en-US/e5e1058d-5a64-4e60-b8e2-0ce327c13058/erroraccessdenied-error-when-trying-to-receive-data-from-bluetooth-low-energy-devices?forum=wdk

More discussion about using `GetFile()` to get a handle to a BLE device:
https://devzone.nordicsemi.com/question/15205/nrf51822-does-not-work-with-windows-81/

Also, here's a collection of code I compiled to "scan" for devices and read characteristics:

https://gist.github.com/programmarchy/c9d02e22d58bfab3f8bb

**Comment by RangerMauve** [comment]
Would it be possible to use something like [NodeRT](https://github.com/NodeRT/NodeRT) to call [the BTLE apis](https://msdn.microsoft.com/en-us/library/windows/hardware/hh450825(v=vs.85).aspx) directly?

**Comment by inf3rno** [comment]
+1

**Comment by sandeepmistry** [comment]
There's some experimental Windows support in the [bluetooth-hci-socket](https://github.com/sandeepmistry/noble/tree/bluetooth-hci-socket) branch.

It uses WinUSB to bypass the Windows Bluetooth stack and talk directly to the Bluetooth 4.0 USB adapter. This layer is handled by [node-bluetooth-hci-socket](https://github.com/sandeepmistry/node-bluetooth-hci-socket). You'll need to use [Zadig tool](http://zadig.akeo.ie/) to swap out the default driver for your adapter, doing so will make it unavailable to Windows. More info, including currently supported adapters can be found [here](https://github.com/sandeepmistry/node-bluetooth-hci-socket#windows)

Inspired by @huysentruitw's [win-beacon](https://github.com/huysentruitw/win-beacon) project!

**Comment by johneke** [comment]
AWESOME!!! +1

**Comment by huysentruitw** [comment]
I'm always happy to inspire :)

**Comment by sandeepmistry** [comment]
@huysentruitw :)

Similar Windows support is also available in the [bleno bluetooth-hci-socket branch](https://github.com/sandeepmistry/bleno/tree/bluetooth-hci-socket), if anyone wants to create a BLE Peripheral on Windows.

**Comment by inf3rno** [comment]
@sandeepmistry Thanks! I'll check it out.

**Comment by sandeepmistry** [comment]
Initial Windows support via WinUSB has been merged into master.

See here for more info: https://github.com/sandeepmistry/noble#windows

Alternative methods to WinUSB might be introduced later on, depending on how the native Windows Bluetooth API's evolve. Controlling the Bluetooth 4.0 USB adapter directly means no comprises to the noble API. However the adapter cannot be used from Windows, when the WinUSB driver has been setup for it.

**Comment by aruneshchandra** [comment]
@sandeepmistry have you looked at the following to see if noble can work more seamlessly Windows 10 ?

BLE Overview
https://msdn.microsoft.com/en-us/library/windows/hardware/jj159880%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396
Bluetooth Advertisement namespace:
https://msdn.microsoft.com/en-us/library/windows.devices.bluetooth.advertisement.aspx
Advertisement Watcher/Publisher Sample:
https://github.com/Microsoft/Windows-universal-samples/tree/master/bluetoothadvertisement
GATT:
https://msdn.microsoft.com/en-us/library/windows/apps/xaml/dn264587.aspx
GATT namespace:
https://msdn.microsoft.com/en-us/library/windows/apps/windows.devices.bluetooth.genericattributeprofile.aspx
LE Device Class:
https://msdn.microsoft.com/en-us/library/windows/apps/windows.devices.bluetooth.bluetoothledevice.aspx

**Comment by sandeepmistry** [comment]
@aruneshchandra yes, I did look at those documentation links. The new Advertisement Watcher API is a step in the right direction.

However, I didn't see any API's or examples to programmatically connect and disconnect to a peripheral once discovered. I'm happy to take a closer look if you have a link or code to do the following programmatically:

1) Start scanning for peripherals
2) Connect to a peripheral
3) Discover services, characteristics, and descriptors of the connected peripheral
4) Interact with characteristics (read, write, subscribe/unsubscribe for notifications)
5) Disconnect

**Comment by aruneshchandra** [comment]
@sandeepmistry thanx for that enumeration and also your continued interest in brining this support to Windows. Looks like there still exist some API gaps to be able to do what you have listed above. 
I will let you know when we have made progress towards closing that gap.

**Comment by sandeepmistry** [comment]
@aruneshchandra thanks!

**Comment by contactsunilkumar** [comment]
Hi, I have tried the "Advertisement Watcher/Publisher Sample" program on Windows 10 desktop, am able to discover the device (say iphone running with advertiser), but not able to connect through BLE and transmit data across.
Just wanted to check, if that support is there to share data across BLE supported devices, or its only the discovery feature as of now with Windows 10.

**Comment by kpathakota** [comment]
@contactsunilkumar There's currently no feature for connecting over BLE, only discovery is supported.

**Comment by contactsunilkumar** [comment]
Hi Kiran

Thanks for your reply.
So currently with windows 10, we can detect the device/beacon using
bluetoothleadvertiserwatcher through compnayid, manufacturing data, but
can't connect to device/beacon to transmit data across.

What about if we make windows 10 machine as advertiser, will the BLEwatcher
running on the other device say iphone will be able to connect and transmit
data or those API'S also are currently un - supported in windows 10.

We are working towards a solution where we need to communicate back n forth
to windows 10 machine with iphone device using BLE.

Regards
Sunil

**Comment by aruneshchandra** [comment]
November update for Windows 10 has the following support. 
* Scanning (Start/Stop) – supported (DeviceWatcher APIs only in foreground)
* Peripheral (Connect, disconnect) – supported
* Connected RSSI – supported
* Characteristic and Descriptor Read/Write – supported
* Characteristic Broadcast/Notify - supported
* Radio states - unknown" | "unsupported" | "poweredOff" | "poweredOn" –  supported

**Samples**
* [Combination of DeviceEnumeration APIs for Discovery and Pairing](https://github.com/Microsoft/Windows-universal-samples/blob/master/Samples/DeviceEnumerationAndPairing/cs/Scenario9_CustomPairDevice.xaml.cs)
* [GATT Client APIs](https://msdn.microsoft.com/en-us/library/windows/apps/windows.devices.bluetooth.genericattributeprofile.aspx) interacting with an already paired device

@sandeepmistry - Do you think there are still gaps that need to be filled?

**Comment by wind-rider** [comment]
@aruneshchandra @sandeepmistry There is GATT sample code for Win10 on this blog http://www.drjukka.com/blog/wordpress/?p=167

**Comment by mhle** [comment]
@aruneshchandra thanks for posting the updates, are there any plans to allow for programmatically connecting to peripherals without first pairing the peripheral with Windows as described in http://www.drjukka.com/blog/wordpress/?p=167

**Comment by timfish** [comment]
@mhle It does look like only 'discoverability' works without pairing currently. They're not being very vocal about their roadmap either. The API isn't great but is still potentially simpler than the `bluetooth-hci-socket` approach for Windows 10.

I'm trying to get my BLE hardware accessible in `atom/electron` on Windows. Currently I would have to ship the hardware with a specific Bluetooth adaptor and apply the WinUSB driver programmatically. There's lots to go wrong there including breaking the customers Bluetooth if they're already using the same adaptor. If I can support Windows 10 without `bluetooth-hci-socket` that would be great! 

I can always open the Bluetooth pairing dialog by running `explorer ms-settings:bluetooth`

**Comment by sandeepmistry** [comment]
@aruneshchandra this is awesome progress! I'll try out the new API's in the next few weeks.

> Do you think there are still gaps that need to be filled?

I have the same concerns as @mhle and @timfish regarding pairing. There might be lots of BLE peripherals in the field which do not need or support pairing. Roughly 90% of the peripherals I've tested with do not require pairing.

The BLE stacks for iOS, OS X, and Android generally do not start the pairing process until a characteristic or descriptor that is "secure" returns an authentication or authorized error code for a GATT request (read, write).

**Comment by trmck** [comment]
@mhle @sandeepmistry @timfish I am currently talking with the DevPlatform team to confirm if the ability to connect without pairing is/will be supported.

**Comment by trmck** [comment]
@mhle @sandeepmistry @timfish Connecting to devices without Bluetooth pairing will be supported for Bluetooth classic later this year (2016) and Bluetooth LE early next year (2017)

**Comment by timfish** [comment]
@trmck Thanks for the update. I've been playing with a BLE prototype today and can only pair in Windows 10 if pairing has been enabled.

@sandeepmistry 
> Roughly 90% of the peripherals I've tested with do not require pairing.

I wonder how many devices don't support pairing? It does complicate the firmware as the pairing keys have to be saved in persistent storage.

**Comment by sandeepmistry** [comment]
@trmck thanks for looking into our pairing question!

@timfish 

> I wonder how many devices don't support pairing? It does complicate the firmware as the pairing keys have to be saved in persistent storage.

Good question. I agree it does complicate the firmware, especially if the peripheral doesn't have any "secure" characteristics.

**Comment by igoutas** [comment]
Hi guys, i am trying to to a BLE pairing with windows 10 using c++. Everything works fine in Windows 7 but I can't do it on Win10. I am using Windows Bluetooth Functions such as BluetoothAuthenticateDeviceEx. Any ideas?

**Comment by timfish** [comment]
I'm not at a computer, but off the top of my head I think the Windows 10
APIs are new and not the same as the pre Win10 ones...

**Comment by igoutas** [comment]
Sure, what I am asking is whether is it supported by windows 10, pairing programmatically based on your experience.

**Comment by sandeepmistry** [comment]
@hgwood has made some great progress on the "Universal Windows Platform" API's and noble!

 https://github.com/hgwood/winble

cc: @aruneshchandra @trmck

**Comment by trmck** [comment]
@sandeepmistry thanks for the link...I'll check it out this evening

**Comment by timfish** [comment]
I currently can't use noble at all on Windows 10 even via WinUSB due to [this bug in node-usb](https://github.com/nonolith/node-usb/issues/109). I'm attempting to get it working in electron shell which requires that native modules be recompiled against the electron-specific node version.

@sandeepmistry looks like winble only runs on Chakra. Will it be simple enough to port to run on node?

**Comment by RangerMauve** [comment]
@timfish If by UWP you mean [this](https://www.npmjs.com/package/uwp), then that requires the Chakra build of Node.js to run.

**Comment by munyirik** [comment]
It's also possible to access UWP APIs without using the node-uwp package. And that way the support isn't only for Chakra. I have an example of how to do that [here](https://github.com/munyirik/win-io/commit/21f499c7701a45fcf50bd10dcb77d78380f79313) and it will involve writing a native addon.

**Comment by hgwood** [comment]
@munyirik Interesting! I wasn't aware it was possible to access UWP APIs without Chakra. I don't have the skills to write native addons (and I'm not looking to acquire those skills), but if at some point there is a package that enables UWP on vanilla Node, I'd be more than happy to replace node-uwp by it in winble.

**Comment by timfish** [comment]
I've avoided writing native modules in many cases by using `node-ffi`. 
There's some overhead in the calls but you can call into native libraries without writing any native code. I don't know how viable it is to use it with `Windows.Devices.Bluetooth.dll` though.
